Mathlib/Logic/Relation.lean

```diff
@@ -0,0 +1,641 @@
+    let ⟨a', he⟩ := dc h
+    ⟨a', he.substr (p := λ x => rβ x (f a)) h, he⟩
+
+end Fibration
+
+/-- The map of a relation `r` through a pair of functions pushes the
+relation to the codomains of the functions.  The resulting relation is
+defined by having pairs of terms related if they have preimages
+related by `r`.
+-/
+protected def Map (r : α → β → Prop) (f : α → γ) (g : β → δ) : γ → δ → Prop := fun c d =>
+  ∃ a b, r a b ∧ f a = c ∧ g b = d
+
+variable {r : α → α → Prop} {a b c d : α}
+
+/-- `ReflTransGen r`: reflexive transitive closure of `r` -/
+@[mk_iff Relation.ReflTransGen.cases_tail_iff]
+inductive ReflTransGen (r : α → α → Prop) (a : α) : α → Prop
+  | refl : ReflTransGen r a a
+  | tail {b c} : ReflTransGen r a b → r b c → ReflTransGen r a c
+
+attribute [refl] ReflTransGen.refl
+
+/-- `ReflGen r`: reflexive closure of `r` -/
+@[mk_iff Relation.refl_gen_iff]
+inductive ReflGen (r : α → α → Prop) (a : α) : α → Prop
+  | refl : ReflGen r a a
+  | single {b} : r a b → ReflGen r a b
+
+/-- `trans_gen r`: transitive closure of `r` -/
+@[mk_iff Relation.trans_gen_iff]
+inductive TransGen (r : α → α → Prop) (a : α) : α → Prop
+  | single {b} : r a b → TransGen r a b
+  | tail {b c} : TransGen r a b → r b c → TransGen r a c
+
+attribute [refl] ReflGen.refl
+
+namespace ReflGen
+
+theorem to_refl_trans_gen : ∀ {a b}, ReflGen r a b → ReflTransGen r a b
+  | a, _, refl => by rfl
+  | a, b, single h => ReflTransGen.tail ReflTransGen.refl h
+
+theorem mono {p : α → α → Prop} (hp : ∀ a b, r a b → p a b) : ∀ {a b}, ReflGen r a b → ReflGen p a b
+  | a, _, ReflGen.refl => by rfl
+  | a, b, single h => single (hp a b h)
+
+instance : IsRefl α (ReflGen r) :=
+  ⟨@refl α r⟩
+
+end ReflGen
+
+namespace ReflTransGen
+
+@[trans]
+theorem trans (hab : ReflTransGen r a b) (hbc : ReflTransGen r b c) : ReflTransGen r a c := by
+  induction hbc
+  case refl => assumption
+  case tail c d _ hcd hac => exact hac.tail hcd
+
+theorem single (hab : r a b) : ReflTransGen r a b :=
+  refl.tail hab
+
+theorem head (hab : r a b) (hbc : ReflTransGen r b c) : ReflTransGen r a c := by
+  induction hbc
+  case refl => exact refl.tail hab
+  case tail c d _ hcd hac => exact hac.tail hcd
+
+theorem symmetric (h : Symmetric r) : Symmetric (ReflTransGen r) := by
+  intro x y h
+  induction' h with z w a b c
+  · rfl
+
+  · apply Relation.ReflTransGen.head (h b) c
+
+
+theorem cases_tail : ReflTransGen r a b → b = a ∨ ∃ c, ReflTransGen r a c ∧ r c b :=
+  (cases_tail_iff r a b).1
+
+@[elab_as_elim]
+theorem head_induction_on {P : ∀ a : α, ReflTransGen r a b → Prop} {a : α} (h : ReflTransGen r a b)
+    (refl : P b refl)
+    (head : ∀ {a c} (h' : r a c) (h : ReflTransGen r c b), P c h → P a (h.head h')) : P a h := by
+  induction h
+  case refl => exact refl
+  case tail b c _ hbc ih =>
+  -- remark: Lean 3 figured out the motive and `apply ih` worked
+  refine @ih (λ {a : α} (hab : ReflTransGen r a b) => P a (ReflTransGen.tail hab hbc)) ?_ ?_;
```

Comment:
:-(

This makes me sad.

Reply:
Yeah I'll try and minimise

Reply:
I reported it as https://github.com/leanprover/lean4/issues/1841